### DISCUSSION TITLE
How to send the information data of memory address between processes?

### DISCUSSION THREAD

**Discussion by givesun** [eclipse-iceoryx/iceoryx#1948] [Q&A]
Hello, I'm going through the source code of posh.
I have few questions.

1) I understood the concept of zero copy in iceoryx.
 The huge real data will be stored in shared memory, and Producer will send the only information for memory address to Consumers.
 So There is no actual data copy. But I want to know how to send or receive the address data between processes?
 Regarding this point, Is there any other ipc is being used? For example Unix domain socket or semaphore? 

2) Regarding lockfree queue.
 As I know the lockfree-queue is being used for zero-copy.
 If there are 3 modules(roudi-daemon, publisher, subscriber), Each modules has its own lockfree-queue?

**Comment by saftabrashid** [comment]
Hi @givesun , did you find your answer? I am also looking for the same info?

**Comment by elBoberido** [comment]
@givesun @saftabrashid 

1.  Is there any other ipc is being used? For example Unix domain socket or semaphore?
    - semaphores are used for event notification
2. Each modules has its own lockfree-queue?
    - there is one lockfree-queue for each subscriber, which means the multiple subscriber will write to the same subscriber queue

**Reply by saftabrashid** [reply]
Thanks a lot for the response. 

semaphores are used for event notification

You mean the consumer is notified through semaphore that the data is written in the memory? I am interested in understanding how does the sharing of (shared memory address) happens between the producer and the consumer. How does the consumer know that the data it has to read is on "this" address?

**Reply by elBoberido** [reply]
@saftabrashid I don't have time to go into details but it's a relocatable data type, which means we just send offsets and those are translated to actual addresses. And yes, the consumer is notified through semaphores.

**Reply by saftabrashid** [reply]
Again, thanks a lot for the response. If I want to get in-depth info on this is there a source? Even if I need to look into the source code, which component should I look at?

**Reply by elBoberido** [reply]
In general you can start here `iceoryx/tree/main/doc/design` and then whatever is in the `iceoryx_posh/.../popo` folders.
